src/controllers/plugins/LuaApi.cpp

```diff
@@ -0,0 +1,288 @@
+#    include "controllers/plugins/PluginController.hpp"
+#    include "messages/MessageBuilder.hpp"
+#    include "providers/twitch/TwitchIrcServer.hpp"
+
+// lua stuff
+#    include "lauxlib.h"
+#    include "lua.h"
+#    include "lualib.h"
+
+#    include <QFileInfo>
+#    include <QLoggingCategory>
+#    include <QTextCodec>
+
+// NOLINTBEGIN(*vararg)
+// luaL_error is a c-style vararg function, this makes clang-tidy not dislike it so much
+namespace chatterino::lua::api {
+
+int c2_register_command(lua_State *L)
+{
+    auto *pl = getApp()->plugins->getPluginByStatePtr(L);
+    if (pl == nullptr)
+    {
+        luaL_error(L, "internal error: no plugin");
+        return 0;
+    }
+
+    QString name;
+    if (!lua::peek(L, &name, 1))
+    {
+        luaL_error(L, "cannot get string (1st arg of register_command)");
+        return 0;
+    }
+    if (lua_isnoneornil(L, 2))
+    {
+        luaL_error(L, "missing argument for register_command: function "
+                      "\"pointer\"");
+        return 0;
+    }
+
+    auto callbackSavedName = QString("c2commandcb-%1").arg(name);
+    lua_setfield(L, LUA_REGISTRYINDEX, callbackSavedName.toStdString().c_str());
+    auto ok = pl->registerCommand(name, callbackSavedName);
+
+    // delete both name and callback
+    lua_pop(L, 2);
+
+    lua::push(L, ok);
+    return 1;
+}
+
+int c2_send_msg(lua_State *L)
+{
+    QString text;
+    QString channel;
+    lua::pop(L, &text);
+    lua::pop(L, &channel);
+
+    const auto chn = getApp()->twitch->getChannelOrEmpty(channel);
+    if (chn->isEmpty())
+    {
+        qCDebug(chatterinoLua) << "send_msg: no channel" << channel;
+        lua::push(L, false);
+        return 1;
+    }
+    QString message = text;
+    message = message.replace('\n', ' ');
+    QString outText = getApp()->commands->execCommand(message, chn, false);
+    chn->sendMessage(outText);
+    lua::push(L, true);
+    return 1;
+}
+
+int c2_system_msg(lua_State *L)
+{
+    if (lua_gettop(L) != 2)
+    {
+        qCDebug(chatterinoLua) << "system_msg: need 2 args";
+        luaL_error(L, "need exactly 2 arguments");
+        lua::push(L, false);
+        return 1;
+    }
+    QString channel;
+    QString text;
+    lua::pop(L, &text);
+    lua::pop(L, &channel);
+    const auto chn = getApp()->twitch->getChannelOrEmpty(channel);
+    if (chn->isEmpty())
+    {
+        qCDebug(chatterinoLua) << "system_msg: no channel" << channel;
+        lua::push(L, false);
+        return 1;
+    }
+    qCDebug(chatterinoLua) << "system_msg: OK!";
+    chn->addMessage(makeSystemMessage(text));
+    lua::push(L, true);
+    return 1;
+}
+
+namespace {
+    void logHelper(lua_State *L, Plugin *pl, QDebug stream, int argc)
+    {
+        stream.noquote();
+        stream << "[" + pl->codename + ":" + pl->meta.name + "]";
+        for (int i = 1; i <= argc; i++)
+        {
+            stream << lua::toString(L, i);
+        }
+        lua_pop(L, argc);
+    }
+
+}  // namespace
```

Comment:
move this to above the `namespace chatterino::...` stuff, could use a `using chatterino;` inside or just prefix Plugin as needed